pkg/v3/observer.go

```diff
@@ -24,50 +20,57 @@ type Postprocessor interface {
 	PostProcess(context.Context, []ocr2keepers.CheckResult) error
 }
 
-// Runner2 is the interface for an object that should determine eligibility state
-type Runner2 interface {
+// Runner is the interface for an object that should determine eligibility state
+type Runner interface {
 	// CheckUpkeeps has an input of upkeeps with unknown state and an output of upkeeps with known state
 	CheckUpkeeps(context.Context, []ocr2keepers.UpkeepPayload) ([]ocr2keepers.CheckResult, error)
 }
 
 type Observer struct {
 	Preprocessors []Preprocessor
 	Postprocessor Postprocessor
-	Runner        Runner2
+	Runner        Runner
 }
 
 // NewObserver creates a new Observer with the given pre-processors, post-processor, and runner
-func NewObserver(preprocessors []Preprocessor, postprocessor Postprocessor, runner Runner2) Observer {
-	return Observer{
+func NewObserver(preprocessors []Preprocessor, postprocessor Postprocessor, runner Runner) *Observer {
+	return &Observer{
 		Preprocessors: preprocessors,
 		Postprocessor: postprocessor,
 		Runner:        runner,
 	}
 }
 
 // Process - receives a tick and runs it through the eligibility pipeline. Calls all pre-processors, runs the check pipeline, and calls the post-processor.
-func (o *Observer) Process(ctx context.Context, tick Tick) error {
+func (o *Observer) Process(ctx context.Context, tick tickers.Tick) error {
 	// Get upkeeps from tick
-	upkeeps, err := tick.GetUpkeeps(ctx)
+	upkeepPayloads, err := tick.GetUpkeeps(ctx)
 	if err != nil {
 		return err
 	}
 
-	var upkeepPayloads []ocr2keepers.UpkeepPayload
 	// Run pre-processors
 	for _, preprocessor := range o.Preprocessors {
-		upkeepPayloads, err = preprocessor.PreProcess(ctx, upkeeps)
+		upkeepPayloads, err = preprocessor.PreProcess(ctx, upkeepPayloads)
 		if err != nil {
 			return err
 		}
 	}
 
+	results := make([]ocr2keepers.CheckResult, 0)
+
 	// Run check pipeline
-	results, err := o.Runner.CheckUpkeeps(ctx, upkeepPayloads)
-	if err != nil {
-		return err
+	if len(upkeepPayloads) > 0 {
+		results, err = o.Runner.CheckUpkeeps(ctx, upkeepPayloads)
```

Comment:
Instead of checking if the length of the payloads before checking upkeeps, can't we just have CheckUpkeeps return early if it receives an empty list?

Reply:
Works for me. Makes the code simpler